src/utils/ucc_coll_utils.c

```diff
@@ -224,11 +235,29 @@ ucc_ep_map_t ucc_ep_map_from_array(ucc_rank_t **array, ucc_rank_t size,
     } else {
         map.type            = UCC_EP_MAP_ARRAY;
         map.array.map       = (void *)(*array);
-        map.array.elem_size = sizeof(ucc_rank_t);
+        map.array.elem_size = elem_size;
     }
     return map;
 }
 
+ucc_ep_map_t ucc_ep_map_from_array(ucc_rank_t **array, ucc_rank_t size,
+                                   ucc_rank_t full_size, int need_free)
+{
+	size_t elem_size = sizeof(ucc_rank_t);
+    return ucc_ep_map_from_array_generic((void **) array, size, full_size,
+                                         need_free, elem_size,
+                                         &ucc_ep_map_get_elem);
```

Comment:
@shimmybalsam i worry that doing that with fn pointer adds a high cost of fn call. This function (get_elem) is called O(team_size) times, ie looping through all the ranks in a team. I think having a branch is much cheaper in this case. So, i would declare 1 ucc_ep_map_get_elem function wich takes a flag is64:
static inline int64 ucc_ep_map_get_elem_function(void **array, int i, int is64)
{
    if (!is64) {
        return (int64_t) (*(ucc_rank_t **)(array))[i];
    } else {
        return (int64_t) (*(uint64_t **)(array))[i];
    }
}

Then ucc_ep_map_from_array_generic would also have is64 argument instead of fn pointer.

Reply:
done